runtime.md

```diff
@@ -1,5 +1,47 @@
 # Runtime and Lifecycle
 
+## State
+
+The runtime state for a container is persisted on disk so that external tools can consume and act on this information.
+The runtime state is stored in a JSON encoded file.
+It is recommended that this file is stored in a temporary filesystem so that it can be removed on a system reboot.
+On Linux based systems the state information should be stored in `/run/oci/containers`.
+The directory structure for a container is `/run/oci/containers/<containerID>/state.json`.
+By providing a default location that container state is stored external applications can find all containers running on a system.
+
+* **version** (string) Version of the OCI specification used when creating the container.
+* **id** (string) ID is the container's ID.
+* **pid** (int) Pid is the ID of the main process within the container.
+* **root** (string) Root is the path to the container's bundle directory.
+
+The ID is provided in the state because hooks will be executed with the state as the payload.
+This allows the hook to perform clean and teardown logic after the runtime destroys it's own state.
+The root directory to the bundle is provided in the state so that consumers can find the container's configuration and rootfs.
+
+*Example*
+
+```json
+{
+    "id": "oci-container",
+    "pid": 4422,
+    "root": "/containers/redis"
+}
+```
+
+Linux systems add some platform specific information to the state.
+
+* **externalFds** Paths to the container's open file descriptors.
+
+*Example Linux*
+
+```json
+{
+    "externalFds": [
+        "/proc/33/fd/1"
```

Comment:
> So the idea is that externalFds[i] tells you where to connect the restoring runtime's FD i+2.

... where to connect the restoring runtime's FD [i].

> That doesn't cleanly handle cases where we have gaps in the file descriptors

The libcontainer interface doesn't assume gaps in the file descriptors
https://github.com/golang/go/blob/92390e47d81713d13b0d9890bf8b153f142a5b8a/src/os/exec/exec.go#L82

> 

> Is “pipe:[18132]” the read or write end of that pipe?

It doesn't matter. We always can reopen it via /proc.

> What if the client is duping descriptors?

It's the main problem which we solved with help of externalFds. We saves uniq id for each file object, so on restore we can find all descriptors which were linked with this file object.

> What if we pass a Unix socket in, and then (after start-time), pass *another* file descriptor down that socket with sendmsg?  What if the client has a listening socket and has accepted a new connection?

If criu found sockets from external net namespaces and they are not marked as external, it will fail. You will know that this container can't be dumped or migrated and this operation doesn't affect the container state.
You are right. It doesn't work for all cases. 

> Linking to CRIU docs would be fine for the parts of this that are covered in the CRIU docs.

http://www.criu.org/Inheriting_FDs_on_restore
http://criu.org/External_UNIX_socket